### PR TITLE
First PoC for UI for customizing profile

### PR DESCRIPTION
**Pull request by Phyks**
See #223 

Very first PoC, still quite some work to do on UI etc.

Screenshot_2019-08-20 Client web BRouter

What do you think about this?

**EDIT**: TODO

- [x] boolean variables should either be checkboxes or yes/no radio
   * probably parsable if default value is true or false or line including comment has
        ```
        1 # set to 0
        0 # set to 1
        ```

### CODE REVIEW

**Comment by nrenner** [comment]
Some thoughts:

- boolean variables should either be checkboxes or yes/no radio
  - probably parsable if default value is `true` or `false` or line including comment has
    - `1   # set to 0`
    - `0   # set to 1`
- internal flags like `validForBikes` probably can be excluded
- this should get its own tab, separate from the editor
  - I added the sidebar to get full height for the editor
  - more user friendly, no need to confront user with profile code and editor UI when setting parameters
- turnInstructionMode would really make sense in the export dialog (in addition?)
  - generally I wouldn't mind having hard-coded inputs for parameters that are also used in server code (outside of profile code)
  - but context-aware: only for supported formats and when currently selected profile has that variable (?)

**Comment by Phyks** [comment]
> internal flags like validForBikes probably can be excluded

Sure, but this would require either maintaining a separate list of such internal variables or having a different notation for these in the BRouter profile.

> this should get its own tab, separate from the editor 

My idea was to have this as an "easy interface" for editing a profile and then a "toggle expert mode" link which would show the current textarea instead.


My goal with this would be, at least in a first time, to get rid of the many variants of some profiles. For instance, the many "trekking" alternatives which actually only differs by a few variables.

**Comment by nrenner** [comment]
> Sure, but this would require either maintaining a separate list of such internal variables 

Yes, why not. Or ignore variables without comment. I guess it doesn't make sense to turn off `validFor*` modes, so why present it to the user? 

> My idea was to have this as an "easy interface" for editing a profile and then a "toggle expert mode" link which would show the current textarea instead.

Just to make sure we are talking about the same thing: 

This sounds like you plan to automatically edit the profile text and replace the default values with the inputs in the client and send it (via "upload" button?) as a customized profile to the server (without any changes in the server)?

What I would like to have is sending the inputs as URL parameters to the server where they override the default values after parsing (don't know where exactly this is/should be done in the server).

**Comment by Phyks** [comment]
> Yes, why not. Or ignore variables without comment. I guess it doesn't make sense to turn off validFor* modes, so why present it to the user?

Indeed, but this would also mean we lose variables such as [`uphillcost`](https://github.com/abrensch/brouter/blob/master/misc/profiles2/trekking.brf#L25).


> This sounds like you plan to automatically edit the profile text and replace the default values with the inputs in the client and send it (via "upload" button?) as a customized profile to the server (without any changes in the server)?

This is the quick and easy going way.

> What I would like to have is sending the inputs as URL parameters to the server where they override the default values after parsing (don't know where exactly this is/should be done in the server).

I agree this would be definitely better. I can look into updating the backend (server) code as well.

**Comment by Phyks** [comment]
Another question on this is what we should do with the BRouter global variables defined in https://github.com/abrensch/brouter/blob/master/brouter-core/src/main/java/btools/router/RoutingContext.java#L127-L188. Most of the time, the default value is not even listed in the profiles.

1. Should the profiles include them with the default values and a `%` comment when they can be of use (`totalMass` etc for instance for bikes profiles)?
2. Should we somehow import them in BRouter-web, no matter the profile?

**Comment by Phyks** [comment]
@nrenner I gave a try at an extended Locus syntax. This PR is updated accordingly. The updated `trekking.brf` profile is available at https://gist.github.com/Phyks/fb9136bdfc81124540ebfcb067048bec (all relevant variables are commented for Locus / BRouter-web).

Current UI is:

Screenshot_2019-09-02 BRouter web client

It still requires a bit of work but if we agree on the backend (profiles) syntax, I'll finish the editing and write a PR to BRouter in parallel of working on the UI.

**Comment by bagage** [comment]
This is looking good, keep going @Phyks :)! I'd just want to put emphasis on [@zossebart comment](https://github.com/nrenner/brouter-web/issues/223#issuecomment-526832098) about translations: maybe the simplest solution is to use the variable name as translation key, eg `ignore_cycleroutes.description` would match `Set to true for better elevation results` and so on?

**Comment by nrenner** [comment]
Thanks for the update, nice progress!

Wanted to say the same about translations, we could also leave off the display name (you already did?) and set it in translations only, like we do for profile names.

**Comment by Phyks** [comment]
> maybe the simplest solution is to use the variable name as translation key, eg ignore_cycleroutes.description would match Set to true for better elevation results and so on?

What about having name and description as currently done in https://github.com/Phyks/brouter/blob/fc13c5c1594575cc92b912536c7fa9b77bd72d1c/misc/profiles2/trekking.brf#L12 with English (just as currently done in BRouter). Apps would be free to use either the provided name and description or use the name key for translations.

In BRouter-web, this would mean, taking https://github.com/Phyks/brouter/blob/fc13c5c1594575cc92b912536c7fa9b77bd72d1c/misc/profiles2/trekking.brf#L12 as an example,

* If `profileParameters.allow_steps.name` key exists in i18n files, it will be used for the name. Otherwise, `allow_steps` will be shown.
* If `profileParameters.allow_steps.description` key exists in i18n files, it will be used for the description. Otherwise, `Set to false to disallow steps` will be shown.

I'll edit the PR accordingly.

**Comment by nrenner** [comment]
> Another question on this is what we should do with the BRouter global variables defined in https://github.com/abrensch/brouter/blob/master/brouter-core/src/main/java/btools/router/RoutingContext.java#L127-L188. Most of the time, the default value is not even listed in the profiles.

I personally would tend to keep this as user friendly as possible and reduced to a minimum of parameters that are likely to be changed in everyday use / from route to route and are ideally self-explaining.

So I would only use those currently in the profiles, or just a selection of them (e.g. probably without `uphillcost`, etc.).

Unless we either have an expert mode or have groups of parameters (like in your profile or the RoutingContext code) and show only the most important and other groups collapsed (maybe as a second step later).

**Comment by Phyks** [comment]
> So I would only use those currently in the profiles, or just a selection of them (e.g. probably without uphillcost, etc.).
> Unless we either have an expert mode or have groups of parameters (like in your profile or the RoutingContext code) and show only the most important and other groups collapsed (maybe as a second step later).

I added the relevant parameters in my update of the trekking profile (https://github.com/Phyks/brouter/blob/fc13c5c1594575cc92b912536c7fa9b77bd72d1c/misc/profiles2/trekking.brf#L8-L40). This is also to try to improve on https://github.com/abrensch/brouter/issues/184 a bit.

We could have groups in the future, parsed from the sections delimited by comments in the profile for instance. Anyways, if we start having translations of parameters names and descriptions in BRouter-web, we could also have this grouping stored within BRouter-web.

In short, my opinion is the edit profile tab is currently for quite advanced users. So, showing all parameters in the first version would probably not be so bad, since we are moving from an advanced users screen to a semi-advanced users screen :) We might refine this afterwards, for sure.

**Comment by Phyks** [comment]
I updated the PR, the user-friendly edit interface should now be styled and working, as well as the switch to the advanced editor.

Here are a few screenshots:

ui

ui2

If you want to give it a try, you should use this PR on BRouter: https://github.com/abrensch/brouter/pull/189.

Happy to take any feedback on this!

**Comment by bagage** [comment]
I cannot make it work yet, with your PR trekking.brf:

> java.lang.IllegalArgumentException: ParseException at line 68: unknown lookup name: bicycle_road

**Comment by nrenner** [comment]
Looking good!

There are a few issues with profile state management and syncing, but I don't know if this is supposed to be fully working already?

**Comment by Phyks** [comment]
Thanks! 

> I don't know if this is supposed to be fully working already?

It should :)

> There are a few issues with profile state management and syncing

Happy to have more details on this. I tested it but probably missed some cases. Thanks!

**Comment by nrenner** [comment]
The (advanced) editor is not synchronized:
1. uncheck "allow_steps", "Save"
2. "Switch to advanced editor"
   -> `assign allow_steps = true` (should be false)
3. "Switch to basic editor"
  -> "allow_steps" is checked again (should be unchecked)

The JavaScript error I see in the console when switching profile and back might be caused by this.

**Comment by Phyks** [comment]
Should be rebased (for conflicts) and fixed.

**Comment by Phyks** [comment]
BRouter counterpart has been merged! :)

**Comment by nrenner** [comment]
Seems like the errors in the console weren't related, I still get them:

`Error getting profile from "profiles2/custom_....brf"` (Profile.js:51)

1. change a parameter in basic editor and "Save"
2. switch profile to "Trekking bike" and back to "Custom"

The request to get the profile text shouldn't be executed for a custom profile, I think this condition shouldn't be true:
https://github.com/nrenner/brouter-web/blob/179f1eedf0430a6a2a5b8499180c43f8d2896347/js/control/Profile.js#L35

There probably needs to be some additional state mangagement for the basic editor.


`Uncaught TypeError: Cannot read property 'replace' of undefined` (Profile.js:110)

3. change a property again and "Save" (after 1. and 2. above)

The same error occurs when app called with a route:

1. http://localhost:3000/#map=18/48.86226/2.33030/osm-mapnik-german_style&lonlats=2.330746,48.86241;2.330974,48.862333
2. change parameter and "Save"

**Comment by Phyks** [comment]
> The same error occurs when app called with a route:

Well caught, this should be fixed now!


> The request to get the profile text shouldn't be executed for a custom profile, I think this condition shouldn't be true:

I agree, but I really don't know what would be the user-friendly behavior here. Current behavior in BRouter-web is not optimal either :/

I pushed a second commit which should fix the issue, I think. Custom profiles are now cached upon upload so that it is easy to recover them!

**Comment by nrenner** [comment]
The fixes look good to me, thanks! Nice idea with the caching.

One thing left is that the profile view/editor is too hidden now in my opinion. You can't see it's there until you scroll all the way down to the end of the options and read the button.

Don't know if users actually use it that way, but I see it not only as an editor for advanced users, but also for viewing the current profile:
- provides a description of the profile in the header comment
- offers some transparency about what the router actually does, 
  e.g. allows to check if and how a specific tag is considered
- shows off the unique feature of a fully customizable profile at runtime

My suggestion would be using [Bootstrap tabs](https://getbootstrap.com/docs/4.3/components/navs/#javascript-behavior) instead of buttons to switch:
- would be the top where you expect navigation and immediately visible
- leaves space for more buttons (plan to add "Save", "Load" and "Share")
- tab UI element implies switching, no need to explain with "Switch to ..."
- maybe name it "Options" and "Profile" instead of "basic editor" and "advanced editor"?
- and rename "Save" to "Apply", because also applicable to "proper" profile parameters later?

What do you think? If you prefer, I could also do this after merging.

**Comment by Phyks** [comment]
Hi,

Thanks for the feedback on this! This sound better in terms of UI indeed.

I'll try to have a look next week, but I'll not have time before Tuesday. So, feel free to merge and edit this if you have time before :)

**Comment by nrenner** [comment]
I'm going to have a look at adding the Bootstrap tabs, unless you already started?

**Comment by Phyks** [comment]
> I'm going to have a look at adding the Bootstrap tabs, unless you already started?

Did not start anything already.

Thanks for the merge and truly sorry for my late reply, I've had far fewer free time than expected lately :/

**Comment by nrenner** [comment]
Implemented as suggested, see #260

**Comment by Phyks** [comment]
Thanks!

**Comment by Stapawe** [comment]
After looking on screenshots, will be in this UI the most important thing - highway costs? Thank you for your great job, especially in the last year or so.
